documentation/ArrayExpressionLoweringToFIR.md

```diff
@@ -0,0 +1,315 @@
+The first observation is that a `fir.loop` is more or less a regular loop and
+it is not returning an ssa value that would be the result of the array expression associated to the loop.
+Instead, a `fir.store` operation in the `fir.loop` must store array result elements into some memory reference.
+This implies that every `fir.loop` computing an array expression must be given a memory reference that can store
+the array result.
+
+
+### Scalar sub-expression must be evaluated at most once.
+In `array + cos(scalar)`, `cos(scalar)` must not be evaluated for every element of `array`.
+
+### Shape may be a compiled time constant shape or not
+In case it is not, FIR must be inserted to compute the `fir.loop` bounds.
+In case the resulting shape is empty (one dimension extent is zero),
+nothing should be evaluated in the expression unless it was evaluated to deduce the shape.
+
+### Short-circuiting requirements
+On top of the null-shape case, they are other cases that should be short-circuited. This is stated here because this
+should combine well with array expression lowering.
+Case identified so far are:
+- `E` is null shaped -> do not evaluate `E`
+- `b.OR.SE`, `b.AND.SE` -> do not evaluate `SE` when `b` is false and `SE` may have side effects.
+- Same as above for the right side if the left side is array shaped?
+- `c1 // c2` if the consumer of the expression only needs a length `<= length(c1)`.
+- Should we skip evaluation on IEEE NaN, Inf... (we know `qNan + foo()` is `qNaN`) ?
+
+This short circuiting logic has to be implemented before the `fir.loop`,
+it may require to produce different `fir.loop` based on the path taken. For instance, in `b.AND.SE`,
+after evaluating `b`, one path should set the temp result to `SE` while the other should set it to `.False.`.
+
+Short-circuiting requires introducing a CFG of basic blocks which makes
+expression lowering changing the basic block context while emitting FIR.
+The creation of basic block for short-circuiting can comply to the following rules:
+
+While lowering expr `E` in basic block `B1`, expression lowering will
+return with the builder pointing at the end of basic block `Bi` such as:
+- `B1` strictly dominates `Bi` (all paths from `B1` leads to `Bi` and all paths arriving to `Bi` went through `B1`).
+- `Bi` may or may not be `B1`.
+- The SSA value returned by expression lowering is defined in `Bi` and is the value of the expression `E`
+- The CFG graph starting at `B1` and ending in `Bi` is a DAG (considering `fir.loop` is an
+ operation and not a CFG loop).
+
+These rules mainly mean that expression lowering does not have to (and should not) create crazy CFG in order
+to achieve short-circuiting.
+
+These rules should guarantee that we do not have to track the CFG state while lowering an expression other than the current insertion point
+(and maybe some places to load symbols. allocate temps...).
+
+### An array shape does not imply an expression is elemental
```

Comment:
Yes, that was a weird sentence.
Rephrased to: `Not all array expressions are elemental expressions`.
The purpose of this trivial observation is just to say that we cannot expect to treat all array expressions as if they were elemental can be transformed into a `fir.loop`.

Reply:
Right, recapping: An array type expression can be composed of subexpressions. Those subexpressions can themselves be of array type or scalar type. The scalar type subexpressions are auto-converted (at least functionally) to array type expressions by projecting the scalar into an array of suitable rank.

But the main point I think you are making here is that it may be possible to short-circuit the entire (array) computation based on the values that need to be projected and how they are used.